### PR TITLE
Improved the performance for multiple _has by exclude previous pids

### PR DESCRIPTION
**Pull request by frankjtao**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java [hunk 1]
```diff
@@ -45,6 +45,8 @@
 import ca.uhn.fhir.jpa.search.builder.predicate.UriPredicateBuilder;
 import ca.uhn.fhir.rest.param.DateRangeParam;
 import ca.uhn.fhir.rest.param.ParamPrefixEnum;
+import jdk.internal.jline.internal.Log;
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java:48]: Was this import accidentally added?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java [hunk 2]
```diff
@@ -542,6 +547,21 @@ public void addResourceIdsPredicate(List<Long> thePidList) {
 		addPredicate(predicate);
 	}
 
+	public void excludeResourceIdsPredicate(Set<ResourcePersistentId> theExsitinghPidSetToExclude) {
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java:550]: It'd probably be good to add a test that specifically tests the functionality this method is adding - Ideally using CaptureQueriesListener (e.g. like the test method `testRead_PidId_AllPartitions()` ) to verify that the generated SQL is doing what you expect.

**Review by jamesagnew — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java (hunk 1 above)
Comment (frankjtao) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java]: it's removed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java (hunk 2 above)
Comment (frankjtao) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java:548]: Will do. Should we reduce this number from 1000 to 500? https://github.com/hapifhir/hapi-fhir/blob/063f4b770014c61187d91edc410314a002609498/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java#L1141 @jamesagnew

**Review by IanMMarshall — APPROVED** [review]
Looks good. Nothing else to add to James' comments.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java:548]: What would be the reason to reduce it?

Comment (frankjtao) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java:548]: When having multiple `_has `in search. e.g. If a patient has 100 Obs, 30 encounter, it meet the _has search criteria,  
e.g.` /Patient?_has:Encounter:subject:class=IMP&_has:Observation:patient:date=gt20200&_has:Observation:patient:code=2345-7`
The query for PID without distinct is 100x100x30. The first query will retrieve 500 same PIDs, then the second query will retrieve 500+1000 = 1500 PIDs before the exclude PID logic kicked in.  Changes the value to 500, the second will only retrieve 1000 PID. It improves the performance, or change to `myMaxResultsToFetch += myDaoconfig.getFetchSizeDefaultMaximum()`, which is 500 default. And it's configurable. @jamesagnew

**Pull request merged**